lib/block_view_stake.go

```diff
@@ -0,0 +1,2496 @@
+	}
+	if currentStakeEntry != nil {
+		bav._deleteStakeEntryMappings(currentStakeEntry)
+	}
+	// 2. Set the PrevStakeEntry.
+	if len(operationData.PrevStakeEntries) < 1 {
+		return fmt.Errorf("_disconnectUnstake: no prev StakeEntry found for %v", currentTxn.PublicKey)
+	}
+	if len(operationData.PrevStakeEntries) > 1 {
+		return fmt.Errorf("_disconnectUnstake: more than one prev StakeEntry found for %v", currentTxn.PublicKey)
+	}
+	bav._setStakeEntryMappings(operationData.PrevStakeEntries[0])
+
+	// Restore the PrevGlobalStakeAmountNanos.
+	bav._setGlobalStakeAmountNanos(operationData.PrevGlobalStakeAmountNanos)
+
+	// Restore the PrevLockedStakeEntry, if exists. The PrevLockedStakeEntry will exist if the
+	// transactor has previously unstaked stake assigned to this validator within the same epoch.
+	// The PrevLockedStakeEntry will not exist otherwise.
+	currentEpochNumber := uint64(0) // TODO: set this
+	// 1. Retrieve the CurrentLockedStakeEntry.
+	currentLockedStakeEntry, err := bav.GetLockedStakeEntry(
+		prevValidatorEntry.ValidatorPKID, transactorPKIDEntry.PKID, currentEpochNumber,
+	)
+	if err != nil {
+		return errors.Wrapf(err, "_disconnectUnstake: ")
+	}
+	// 2. Delete the CurrentLockedStakeEntry.
+	bav._deleteLockedStakeEntryMappings(currentLockedStakeEntry)
+	// 3. Set the PrevLockedStakeEntry, if exists.
+	if len(operationData.PrevLockedStakeEntries) > 1 {
+		return fmt.Errorf("_disconnectUnstake: more than one prev LockedStakeEntry found for %v", currentTxn.PublicKey)
+	}
+	if len(operationData.PrevLockedStakeEntries) == 1 {
+		bav._setLockedStakeEntryMappings(operationData.PrevLockedStakeEntries[0])
+	}
+
+	// Disconnect the basic transfer.
+	return bav._disconnectBasicTransfer(
+		currentTxn, txHash, utxoOpsForTxn[:operationIndex], blockHeight,
+	)
+}
+
+func (bav *UtxoView) _connectUnlockStake(
+	txn *MsgDeSoTxn,
+	txHash *BlockHash,
+	blockHeight uint32,
+	verifySignatures bool,
+) (
+	_totalInput uint64,
+	_totalOutput uint64,
+	_utxoOps []*UtxoOperation,
+	_err error,
+) {
+	// Validate the starting block height.
+	if blockHeight < bav.Params.ForkHeights.ProofOfStakeNewTxnTypesBlockHeight ||
+		blockHeight < bav.Params.ForkHeights.BalanceModelBlockHeight {
+		return 0, 0, nil, RuleErrorProofofStakeTxnBeforeBlockHeight
+	}
+
+	// Validate the txn TxnType.
+	if txn.TxnMeta.GetTxnType() != TxnTypeUnlockStake {
+		return 0, 0, nil, fmt.Errorf(
+			"_connectUnlockStake: called with bad TxnType %s", txn.TxnMeta.GetTxnType().String(),
+		)
+	}
+
+	// Grab the txn metadata.
+	txMeta := txn.TxnMeta.(*UnlockStakeMetadata)
+
+	// Validate the txn metadata.
+	if err := bav.IsValidUnlockStakeMetadata(txn.PublicKey, txMeta); err != nil {
+		return 0, 0, nil, errors.Wrapf(err, "_connectUnlockStake: ")
+	}
+
+	// Convert TransactorPublicKey to TransactorPKID.
+	transactorPKIDEntry := bav.GetPKIDForPublicKey(txn.PublicKey)
+	if transactorPKIDEntry == nil || transactorPKIDEntry.isDeleted {
+		return 0, 0, nil, RuleErrorInvalidStakerPKID
```

Comment:
Let's wrap these rule errors - will help us track which type of txn we hit it in.